### PR TITLE
Implement Context, add reader, repl pacakges

### PR DESCRIPTION
**Pull request by spy16**
Highlights:

1. `Context` is the main type in parens package. `Context` exposes only `Eval()`
2. stack related types/functions are private (See #19 for reasoning).
3. Context has a `fork()` method which `GoExpr` uses to create a child context and launch a goroutine.

### CODE REVIEW

File: reader/errors.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+// Is returns true if the other error is same as the cause of this error.
+func (e Error) Is(other error) bool { return errors.Is(e.Cause, other) }
+
+// Unwrap returns the underlying cause of the error.
+func (e Error) Unwrap() error { return e.Cause }
+
+func (e Error) Error() string {
+	switch {
+	case errors.Is(e.Cause, ErrUnmatchedDelimiter):
```
Comment (spy16) [inline reader/errors.go:45]: just experimenting with a different error handling approach. not really sure which is better.

File: expr.go [hunk 2]
```diff
@@ -73,7 +73,7 @@ func (ife IfExpr) Eval(ctx *Context) (value.Any, error) {
 	return ctx.Eval(ife.Else)
 }
 
-// DoExpr represents the (do expr*) form.
+// DoExpr represents the (do expt*) form.
```
Comment (lthibault) [inline expr.go]: Is `expt` a typo?

**Edit:** I'm guessing not, since it's used in several places.  What does it stand for?

File: parens.go [hunk 3]
```diff
@@ -8,32 +8,30 @@ import (
...
-	ctx := &Context{}
-	for _, opt := range withDefaults(opts) {
-		opt(ctx)
-	}
-
+	ctx := &Context{ctx: context.Background()}
 	ctx.push(stackFrame{
 		Name:          "<global>",
-		ConcurrentMap: ctx.mapFactory(),
+		ConcurrentMap: newMutexMap(),
```
Comment (lthibault) [inline parens.go]: The global stack frame should also use the user-specified `ConcurrentMap` implementation.  To do so, we should apply options to `ctx` first, and then push the `stackFrame` like this:

```go
ctx.push(stackFrame{
    Name:          "<global>",
    ConcurrentMap: ctx.mapFactory(),
})
```

File: parens.go [hunk 4]
```diff
@@ -8,32 +8,30 @@ import (
 
 // New returns a new root context initialised based on given options.
 func New(opts ...Option) *Context {
```
Comment (lthibault) [inline parens.go]: Any particular reason why we are returning a pointer instead of a concrete value?

I tend to use pointers when either:

1. The constructor can return an error, and I want to return a null value (i.e.:  `return nil, err`)
2. The concrete value is large and copying it is likely to hurt performance

Since `New` neither fails nor returns a large value (it pretty much only contains pointer types), I'm inclined to return a concrete `Context`.  The benefits are that it removes an opportunity for nil pointer dereference errors and that it reduces pointer-chasing in a heavily-used structure.

File: value/utils.go [hunk 5]
```diff
@@ -49,7 +48,6 @@ func SeqString(seq Seq, begin, end, sep string) (string, error) {
 	}
 
 	return strings.TrimRight(b.String(), sep) + end, err
-
```
Comment (lthibault) [inline value/utils.go]: Trailing whitespace.

File: cmd/parens/main.go [hunk 6]
```diff
@@ -1,13 +1,23 @@
...
-	ctx := parens.NewContext()
-	fmt.Println(ev.Eval(ctx, "hello"))
+	globals := map[string]value.Any{
+		"pi":      value.Float64(3.1416),
+		"version": &value.String{Value: "1.0"},
+	}
+
+	ctx := parens.New(parens.WithGlobals(globals))
+
+	_ = repl.New(ctx,
```
Comment (lthibault) [inline cmd/parens/main.go]: Minor point:  our use of `ctx` is a bit confusing because it suggests a `context.Context`.  We may want to use another convention , e.g. `cx` or just `c`.

File: expr.go [hunk 7]
```diff
@@ -48,42 +47,42 @@ type DefExpr struct {
 	Value value.Any
 }
 
-// Eval the expression
-func (de DefExpr) Eval(ctx Context, ev Evaluator) (value.Any, error) {
+// Eval creates a global binding in the root context.
```
Comment (lthibault) [inline expr.go]: The "root context" part of the docstring is a bit confusing.  It suggests that `ctx` is the root context, which is not necessarily the case.

How about:  `// Eval creates a symbol binding in the global (root) stack frame.`?

File: parens.go [hunk 8]
```diff
@@ -1,88 +1,156 @@
...
+	}
+	return ctx
 }
 
-// Analyzer implementation can be set on the Evaluator to override evaluation
-// rules for forms. See WithAnalyzer().
-type Analyzer interface {
-	Analyze(ev Evaluator, form value.Any) (Expr, error)
+// EvalAll evaluates each value in the list against the given ctx and returns
+// a list of resultant value.
```
Comment (lthibault) [inline parens.go]: Typo:  
>[...] and returns a list of resultant value**s**.

File: parens.go [hunk 9]
```diff
@@ -1,88 +1,156 @@
...
 	}
 
-	return ev.analyze(form)
+	return ctx.analyzer.Analyze(ctx, form)
 }
 
-func (ev Evaluator) expand(form value.Any) (value.Any, error) {
-	if ev.Expander == nil {
-		return basicExpander{}.Expand(ev, form)
+// fork creates a child context from the parent and returns, which can be
```
Comment (lthibault) [inline parens.go]: I would reword this to something like:

>fork creates a child context from ctx and returns it.  The child context can be used as context for an independent thread of execution.

This clarifies that `ctx` is the parent, and that the child is the return value.

File: parens.go [hunk 10]
```diff
@@ -1,88 +1,156 @@
...
+func (ctx *Context) pop() (frame *stackFrame) {
+	if len(ctx.stack) == 0 {
+		panic("pop from empty stack")
 	}
+	frame, ctx.stack = &ctx.stack[len(ctx.stack)-1], ctx.stack[:len(ctx.stack)-1]
+	return frame
+}
+
+func (ctx *Context) setGlobal(key string, value value.Any) {
+	// TODO: verify this. what is expected if this is a child context?
```
Comment (lthibault) [inline parens.go]: I can't remember if this TODO is from you or from me ... 🤔   Either way, here's my viewpoint:  goroutines share a common, global context with all goroutines.  As such, anyone should be able to (re)bind symbols in the global scope.  In fact, this is why stack frames have a _concurrent_ map.

In other words:

- It's the user's responsibility to synchronize
- Globals are best avoided where possible

Therefore, here's the behavior I expect when calling `setGlobal` from a child context:

1. index position `0` in the stack array
2. perform lookup against the concurrent map
3. concurrent map is shared by `stack[0]` in all contexts, so we're mutating a shared variable.  Corollary:  this change will be visible across all goroutines.

File: reader/errors.go (hunk 1 above)
Comment (lthibault) [reply reader/errors.go:45]: I think I like this approach better.  What I ultimately want out of error handling is this:

1. To be able to test (e.g. by type assertion) that this is a _parens_ error.
2. To know the reader's position in the stream that caused the error
3. Ideally to have access to the form/expr/value that caused the error (e.g. so that I can print or analyze it)
4. Ideally, to have some notion of error _types_ (was it a SyntaxError? An IOError? Something else?).  These error types should be programmatically analyzable at runtime too (e.g. via type assertions)

We don't have to solve all of this at once, but that's what I have in mind.

File: value/atoms.go [hunk 11]
```diff
@@ -9,12 +10,18 @@ import (
 // Nil represents the Value 'nil'.
 type Nil struct{}
 
+// SExpr returns a valid s-expression representing Nil.
+func (Nil) SExpr() (string, error) { return "nil", nil }
+
 // HashCode always returns 0.
 func (n Nil) HashCode() int64 { return 0 }
```
Comment (lthibault) [inline value/atoms.go]: In a [previous thread in the Sabre repo](https://github.com/spy16/sabre/pull/25#issuecomment-636502460), we had discussed the idea that methods like `HashCode` might return a hashable representation of the value, instead of the hash itself.  This would effectively defer hashing to the user, who is then free to use whatever algorithm he sees fit.  Is this still the case here?  I'm still very strongly in favor of this approach.

If this is still the case, why return an `int64` instead of e.g. `[]byte` or even an `interface{}` that the implementer must ensure is hashable?

The reason I ask is that in Wetware, all types are actually implemented on top of `[]byte`.  This was done intentionally because there's a lot of hashing going on, and I was hoping to for e.g. `Vector.HashCode()` to simply return the underlying `[]byte`.

**Review by lthibault — APPROVED** [review]
The high-level design and public API look good to me.  My comments relate to lower-level implementation details.

I'm happy for you to merge this optimistically, and for us to address/fix the issues I've raised in later PRs since this one is getting pretty big.

File: expr.go (hunk 2 above)
Comment (spy16) [reply expr.go]: It is a typo actually 😅 . Will fix that. I might have copied and pasted it if it's in multiple places.

File: parens.go (hunk 3 above)
Comment (spy16) [reply parens.go]: Yea, this was bit of a chicken-and-egg problem 😅 .. I was trying to add a `WithGlobals` option that would set some global vars during init. But if the frame didn't exist/initialised, this would panic. so i had to push a frame with concurrent map initialised. But definitely agreed that all frames should be using the map returned by the set map-factory.

File: parens.go (hunk 4 above)
Comment (spy16) [reply parens.go]: Agreed. I am just bit skeptical of value types sometime since all the edge cases it can cause (https://github.com/golang/go/wiki/CodeReviewComments#receiver-type).. 

But i guess here it would be better to use Value type.

File: value/atoms.go (hunk 11 above)
Comment (spy16) [reply value/atoms.go]: Oh yea, we could go down that route. Actually, the `HashCode` function was something i was experimenting on, in Zulu based on Java spec. This was not meant to be part of parens. I guess we copied the value package from there. so it ended up here.

File: parens.go (hunk 3 above)
Comment (lthibault) [reply parens.go]: How about we do this:

```go
func New(opts ...Option) *Context {
        ctx := &Context{ctx: context.Background()}

	ctx.push(stackFrame{
	        Name:          "<global>",
                ConcurrentMap: newMutexMap(),

    	for _, opt := range withDefaults(opts) {
		opt(ctx)
	}

        m := ctx.mapFactory()
        for k, v := range ctx.stack[0].Native() {
                m.Store(k, v)
        }
        ctx.stack[0].ConcurrentMap = m

	return ctx
}
```

It's a bit ugly, but it's probably Good Enough (TM).

File: value/atoms.go (hunk 11 above)
Comment (lthibault) [reply value/atoms.go]: >I guess we copied the value package from there. so it ended up here.

Yeah, I think that was me.  

Go's [`Hash`](https://golang.org/pkg/hash/#Hash) package works with `[]byte` as input, so I think we should define a `Hashable() []byte` method.

I'm inclined to make this part of the `value.Any` contract as well, since this is going to be required for things like hashmap types.  I'd rather not pay the price of type assertion on something so fundamental.

File: reader/errors.go (hunk 1 above)
Comment (spy16) [reply reader/errors.go:45]: Agreed. That's a fair ask. I will keep this as is for now. I will create an issue for error handling improvements.

File: cmd/parens/main.go (hunk 6 above)
Comment (spy16) [reply cmd/parens/main.go]: Yes, I feel the same too. I am wondering if renaming `Context` to `Env` would help ? or should we keep `Context` and like you said, use `c` perhaps ?

File: parens.go (hunk 10 above)
Comment (spy16) [reply parens.go]: This TODO is from me. Just to confirm we are on the same page, you are saying the current behaviour is correct. Current behaviour is basically:

1. Locate root context (if this the main thread, then `ctx` itself is the root. otherwise, locate the root by following `parent` pointer)
2. Do a `Store()` into the 0th frame of the stack in root context.

File: cmd/parens/main.go (hunk 6 above)
Comment (lthibault) [reply cmd/parens/main.go]: Yeah, good call.  Let's rename it to `Env`, which has the added bonus of being shorter as well.

File: parens.go (hunk 3 above)
Comment (spy16) [reply parens.go]: Yea, I thought of this too (only i was planning to do it in `WithMapFactory` option). wasn't sure if there was a better way.

File: parens.go (hunk 4 above)
Comment (spy16) [reply parens.go]: What do you suggest we do with `Analyzer.Analyze(ctx *Context....)`  and other similar signatures ?

File: parens.go (hunk 10 above)
Comment (lthibault) [reply parens.go]: >you are saying the current behaviour is correct.

Correct. 👍 

>1. Locate root context (if this the main thread, then ctx itself is the root. otherwise, locate the root by following parent pointer)
>2. Do a Store() into the 0th frame of the stack in root context.

Whoops, I had overlooked the fact that we are now storing a pointer to the parent `Context`.  Why are we doing this?
It seems like [we are still copying the full stack](https://github.com/spy16/parens/pull/20/files/bb7704e1c9191c3e1a10b345608c8c5cf2938344#diff-093c70a89dda8ea47ef4bc38df892a5bR86) from the parent to the child context.  Why chase pointers instead of using `ctx.stack[0].Store(...)`?

File: parens.go (hunk 4 above)
Comment (lthibault) [reply parens.go]: I think they should take concrete `Context` values.  In fact, I think everything should take concrete values for `Context` _except_ `parens.Option`, which obviously needs a pointer.

Rationale:  performance (i.e.:  cache locality and inlining).  If this ends up being one of those weird edge cases where value types are slower than pointers, we can (and should!) switch to a pointer.

File: parens.go (hunk 10 above)
Comment (spy16) [reply parens.go]: actually the full copy is wrong.. but pointer chasing would be needed. 

1. Imagine you are currently in the root env `E1` and running some code in thread `T1`. You now launch `T2` which forks `E1` to create `E2`. You now have `T1` and `T2` running concurrently against `E1` and `E2` respectively. 
2. Let's say `T1` modifies a global binding in `E1`. This needs to be reflected in `T2` as well since its a global binding. 

If we took a copy of the stack during fork, this update by T1 would be lost to T2.

File: value/atoms.go (hunk 11 above)
Comment (spy16) [reply value/atoms.go]: Yea, i like the `Hashable() []byte` method. Not sure about adding it to `Any` though (since in our last discussion we decided not to do that since it's not possible in Go to get some hashable value for all value types)

File: parens.go (hunk 4 above)
Comment (spy16) [reply parens.go]: this needs some more thought i think 🤔 . 

1. `Expr` type anyways needs to take a pointer to `Env` since it might need to do push a stack frame etc.  
2. Analyser and Expander, I am not sure if this can be a scenario. let me know if you think this is valid: some Exprs might need to maintain a pointer to the env they were analysed/expanded in so that when they are evaluated, they can generate a side effect in those envs..

File: value/atoms.go (hunk 11 above)
Comment (spy16) [reply value/atoms.go]: I am removing the HashCode from all values for now. Will add an issue for hashable types.

Comment (lthibault) [reply value/atoms.go]: >(since in our last discussion we decided not to do that since it's not possible in Go to get some hashable value for all value types)

Maybe I'm missing your point, but the advantage of a `Hashable()` method is that it gives you a way to produce a hashable representation of types that aren't hashable in native Go.  As an example, let's imagine we've defined a Parens type that satisfies `value.Any`, that is actually a `[]string` under the hood (this example is loosely based on Wetware's `Path` type).

```go
type Path []string

func (p Path) SExpr() (string, error) {
    var b strings.Builder
    b.WriteString("/")
    for _, s := range p {
        b.WriteString(s + "/")
    }
    return b.String(), nil
}
```

Now `[]string` is an unhashable type, but `Hashable()` lets us hash it anyway.  Here's a dead-simple implementation:

```go
func (p Path) Hashable() []byte {
    var b strings.Builder
    b.WriteString("/")
    for _, s := range p {
        b.WriteString(s + "/")
    }
    return b.Bytes()
}
```

🎉 Now you can hash a `[]string`!

Of course, if I've completely missed your point, please let me know 😅

File: parens.go (hunk 10 above)
Comment (lthibault) [reply parens.go]: >If we took a copy of the stack during fork, this update by T1 would be lost to T2.

Hmm, I don't think that's right.

Recall that [`stackFrame` holds a `ConcurrentMap` interface](https://github.com/spy16/parens/blob/feature/interpreter/parens.go#L155), which is a pointer type.  So while we are copying the stack, each stack frame contains a _pointer_ to a concurrent map (again, because interfaces are "just pointers").

https://play.golang.org/p/s9-7k04wl83

File: value/atoms.go (hunk 11 above)
Comment (spy16) [reply value/atoms.go]: []string works great. I'm thinking about the more native values like Func types for example. We should be able to expose go func values or some arbitrary struct values into parens env somehow so that they can be called/used from within lisp code. But if we expect every value type to have hashable, then we need to find a way that consistently returns same bytes for a specific value.. I don't think there's a good way to do that. 

On the other hand, may be we can use the fact that expr and value types are now different and do something innovative 🤔

Comment (lthibault) [reply value/atoms.go]: > I'm thinking about the more native values like Func types for example.

This is effectively the defunctionalization/refunctionalization problem.  The trick is to substitute a symbol in lieu of the function object itself.  Here's a naïve solution for a native go `func()`:

```go
type GoFunc func()

func (f GoFunc) Hashable() []byte {
    return []byte(reflect.ValueOf(f).String())
}
```

The `reflect.Valueof(...).String()` call can be substituted with a lookup table as well.

File: parens.go (hunk 4 above)
Comment (lthibault) [reply parens.go]: >Expr type anyways needs to take a pointer to Env since it might need to do push a stack frame etc.

Ah, right of course.  In that case, we should probably just pass a `*Context` everywhere.

**Comment by lthibault** [comment]
I've taken the liberty of pushing a few changes to this branch, namely:

- Simplify implementation of `String`, `Keyword` and `Symbol` now that they are not caching hash values anymore.
- Remove unused `HashCode` benchmarks
- Add `true`, `false`,  and `nil` to globals in `main.go`
- **Bugfix:**  nil pointer panic on `Env.setGlobal` if `parens.WithGlobal` was not used
- Add unit test for `GoExpr` (in order to verify [my assertions](https://github.com/spy16/parens/pull/20#discussion_r488007526) about stack-copying vs `Env.parent`)

On this final point, please check out #21.  I think we should eschew `Env.parent` in favor of a pure stack-copying approach, but happy to discuss, as always. 😃

File: parens.go (hunk 10 above)
Comment (spy16) [reply parens.go]: I think there's a mistake in this design. Vars in a stack-frame cannot be shared across frames/envs except for frame-0 of root env. For example, consider that we are currently invoking `foo` so there is a corresponding frame in the stack with the local bindings (arguments, let vars etc.). Now `foo` invokes `bar`, so we push a frame and execute `bar`. But key requirement here is that we isolate the frame (i.e., vars from `foo` frame should not be leaked to `bar`).. 

With that in mind, I think we don't need a concurrent-map per frame. We just need a native map and only at Env level we need to maintain a single concurrent map. So we need to change the structure to this:

```go
type Env struct {
    globals ConcurrentMap
    stack []stackFrame
    // .... other fields ....
}
``` 

```go
type stackFrame struct {
    args []Any
    name string
    vars map[string]Any
}
```

But removing the parent pointer should be possible since we can pass the concurrent map across to envs when we `fork()`. 

Resolve would first look at top of the stack and if the binding for a key is not found there, it looks up in `currentEnv.globals` map.

(That also means, we don't need to push a `global` frame into the stack when we init Env. we just keep empty stack and the global frame is implied)

Comment (lthibault) [reply parens.go]: Ah, ok.  I think I see what you mean!  My initial concern was that concurrent read/writes to local stack frames was possible, but I'm beginning to see why this is not possible in Lisp.  If so, I'm 100% onboard -- I would love nothing more than to get native map performance on local frames! 😃 

As a sanity check, and to make sure we're on the same page, consider this contrived Go code:

```go
func main() {
    var x int
    go print(x) // race condition:  may read x while x is incrementing
    x++
}
```

In this example, some sort of synchronization would be needed on `x`.  My concern was that a similar thing could occur in lisp code, where `foo` calls `(go bar)`, and then modifies a foo-local variable that `bar` is also reading/modifying.  If this were possible, we would need a `ConcurrentMap` for local stack frames as well.

Upon further reflection, it seems like this scenario is impossible Lisp due to how local variables are bound.  The reasoning is:

1. Assumption 1:  local variables can only be bound with `let`.
2. Assumption 2:  `let` pushes a new frame onto the stack.

Therefore:  once bound, a local variable cannot be rebound or modified.  

Is this correct?

Comment (lthibault) [reply parens.go]: Addendum:  can we still have an option to set the `ConcurrentMap` implementation?  Map + `sync.RWMutex` is almost always the right choice, but I'd still like the possibility of using a more optimized datastructure if needed.

**Comment by lthibault** [comment]
@spy16 I think this should be good to merge.  Shall we pull the trigger?

**Comment by spy16** [comment]
Yea. Looks good I think. Let me fix up the readme and package docstrings and merge it today.

**Comment by spy16** [comment]
Yea. Looks good I think. Let me fix up the readme and package docstrings and merge it today.

**Comment by spy16** [comment]
Yea. Looks good I think. Let me fix up the readme and package docstrings and merge it today.

File: parens.go (hunk 10 above)
Comment (spy16) [reply parens.go]: More or less. But I don't think it's specific to how Lisp does it. It's usually how scopes work. Even in Go, you couldn't share the bindings map of a stack frame. Consider this example:

```go
func main() {
   foo := 100
   go someFunc()
   foo += 100
}

func someFunc() {
   // this is an error because the stack frame corresponding to invocation of `main`
   // and its variables map are not shared to the stack frame corresponding to this
   // function call (i.e., someFunc).
   fmt.Println(foo) 
}
```

Comment (lthibault) [reply parens.go]: Ah good point -- I was confusing closures with function calls. 👍 

Thanks.  Resolving this now.
